cros_gralloc/cros_gralloc_driver.cc

```diff
@@ -169,40 +175,44 @@ cros_gralloc_driver::cros_gralloc_driver()
 		drmFreeVersion(version);
 	}
 
-	// open the first render node
 	if (availabe_node > 0) {
-		drv_render_ = drv_create(node_fd[0]);
+		if (i915_node_idx != -1) {
+			// Prefer i915 for performance consideration.
+			//
+			// TODO: We might have multiple i915 devices in the system and in
+			// this case we are effectively using the one with largest card id
+			// which is normally the dGPU VF or dGPU passed through device.
```

Comment:
So if BM with iGPU and dGPU, always use the dGPU?
If we boot VM with iGPU SRIOV +dGPU passthrough, always use the dGPU passthrough?
Is it a requirement or new define for the gpu usage?

Reply:
If we need to scan out on the dGPU display port, I am afraid that we have no choice other than allocating the buffers on dGPU local memory. As for iGPU display, framebuffers can migrate to system memory automatically when being imported into virtio-GPU.

Overall, if only one allocator is available, dGPU might be better since it can cover the use case of iGPU.

Reply:
In mesa, we detect whether the gpu is dgpu by check if it has local memory.
I only afraid it will change the usage scenario so far.
If we boot VM1 with iGPU sriov +dGPU passthrough, before we use the igpu to allocate buffer, dGPU could used for render.
with your change, the VM1 is same as VM3?

Reply:
This change won't change the way renderer device is selected in Mesa, but I do think that we need a patch in Mesa to choose dGPU as default EGL renderer...